module/VuFind/src/VuFind/DeveloperSettings/DeveloperSettingsService.php

```diff
@@ -0,0 +1,222 @@
+    protected int $keyLimitPerUser;
+
+    /**
+     * Update interval to update last_used values in minutes.
+     *
+     * @var int
+     */
+    protected int $updateInterval = 60;
+
+    /**
+     * Constructor.
+     *
+     * @param ApiKeyServiceInterface $apiKeyService  API key database service
+     * @param array                  $apiKeySettings Section API_Keys from main configuration.
+     */
+    public function __construct(
+        protected ApiKeyServiceInterface $apiKeyService,
+        protected array $apiKeySettings
+    ) {
+        $this->keyLimitPerUser = $apiKeySettings['key_limit'] ?? 5;
+    }
+
+    /**
+     * Generate a new api key token
+     *
+     * @param UserEntityInterface $user User to create salt for
+     *
+     * @return string
+     */
+    protected function createNewToken(UserEntityInterface $user): string
+    {
+        $salt = $this->apiKeySettings['token_salt'] ?? null;
+        if (!$salt || strlen($salt) < 10) {
+            throw new \Exception('DeveloperSettingsService: Invalid token_salt provided');
+        }
+        $valuesForToken = [
+            $user->getEmail(),
+            $user->getFirstname(),
+            $user->getLastname(),
+            time(),
+            $salt,
+        ];
+        return hash('sha256', implode('|', $valuesForToken));
+    }
+
+    /**
+     * Retrieve API keys for user.
+     *
+     * @param UserEntityInterface $user User
+     *
+     * @return ApiKeyEntityInterface[]
+     */
+    public function getApiKeysForUser(UserEntityInterface $user): array
+    {
+        return $this->apiKeyService->getApiKeysForUser($user);
+    }
+
+    /**
+     * Generate an API key for a user.
+     *
+     * @param UserEntityInterface $user  User
+     * @param string              $title Title for the API key
+     *
+     * @return ApiKeyEntityInterface|false API key entity on success, false on failure.
+     */
+    public function generateApiKeyForUser(UserEntityInterface $user, string $title): ApiKeyEntityInterface|false
+    {
+        $tokens = $this->apiKeyService->getApiKeysForUser($user);
+        if ($this->apiKeysBlocked($tokens)) {
+            return false;
+        }
+        // Generate unique id from date and users id.
+        $newKey = $this->apiKeyService->createEntity();
+        $date = new DateTime();
+        $newKey->setToken($this->createNewToken($user))
+            ->setUser($user)
+            ->setCreated($date)
+            ->setLastUsed($date)
+            ->setTitle($title);
+        $this->apiKeyService->persistEntity($newKey);
+        return $newKey;
+    }
+
+    /**
+     * Set the last used value to the API key. By default this will be only updated once every hour.
+     *
+     * @param ApiKeyEntityInterface $apiKey API key
+     *
+     * @return void
+     */
+    protected function updateLastUsed(ApiKeyEntityInterface $apiKey): void
+    {
+        if (time() - $apiKey->getLastUsed()->getTimestamp() < $this->updateInterval * 60) {
```

Comment:
Isn't this reversed? Should be:

```suggestion
        if (time() - $apiKey->getLastUsed()->getTimestamp() >= $this->updateInterval * 60) {
```

Reply:
Actually yes, added a test for this also.